examples/edit_text.rs

```diff
@@ -0,0 +1,509 @@
+                + TEXT_X;
+        let rect = Rect::from_points(
+            Point::new(selection_start_x.min(selection_end_x - 1.0), TEXT_Y),
+            Point::new(
+                selection_end_x.max(selection_start_x + 1.0),
+                TEXT_Y + FONT_SIZE as f64,
+            ),
+        );
+        sb.fill(
+            Fill::NonZero,
+            Affine::IDENTITY,
+            &Brush::Solid(Color::rgba8(0, 0, 255, 100)),
+            None,
+            &rect,
+        );
+        sb.pop_layer();
+    }
+}
+
+impl WinHandler for WindowState {
+    fn connect(&mut self, handle: &WindowHandle) {
+        self.handle = handle.clone();
+        let token = self.handle.add_text_field();
+        self.text_input_token = Some(token);
+        self.handle.set_focused_text_field(Some(token));
+        self.schedule_render();
+    }
+
+    fn prepare_paint(&mut self) {}
+
+    fn paint(&mut self, _: &Region) {
+        self.render();
+        self.schedule_render();
+    }
+
+    fn idle(&mut self, _: IdleToken) {
+        self.render();
+        self.schedule_render();
+    }
+
+    fn size(&mut self, size: Size) {
+        self.size = size;
+    }
+
+    fn request_close(&mut self) {
+        self.handle.close();
+    }
+
+    fn destroy(&mut self) {
+        Application::global().quit()
+    }
+
+    fn acquire_input_lock(
+        &mut self,
+        _token: TextFieldToken,
+        _mutable: bool,
+    ) -> Box<dyn InputHandler> {
+        Box::new(AppInputHandler {
+            state: self.document.clone(),
+            window_size: self.size,
+            window_handle: self.handle.clone(),
+        })
+    }
+
+    fn release_input_lock(&mut self, _token: TextFieldToken) {
+        // no action required; this example is simple enough that this
+        // state is not actually shared.
+    }
+
+    fn key_down(&mut self, event: KeyEvent) -> bool {
+        if event.key == Key::Character("c".to_string()) {
+            // custom hotkey for pressing "c"
+            println!("user pressed c! wow! setting selection to 0");
+
+            // update internal selection state
+            self.document.borrow_mut().selection = Selection::caret(0);
+
+            // notify the OS that we've updated the selection
+            self.handle
+                .update_text_field(self.text_input_token.unwrap(), Event::SelectionChanged);
+
+            // repaint window
+            self.handle.request_anim_frame();
+
+            // return true prevents the keypress event from being handled as text input
+            return true;
+        }
+        false
+    }
+
+    fn as_any(&mut self) -> &mut dyn Any {
+        self
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct ParleyBrush(pub Brush);
+
+impl Default for ParleyBrush {
+    fn default() -> ParleyBrush {
+        ParleyBrush(Brush::Solid(Color::rgb8(0, 0, 0)))
+    }
+}
+
+impl PartialEq<ParleyBrush> for ParleyBrush {
+    fn eq(&self, _other: &ParleyBrush) -> bool {
+        true // FIXME
```

Comment:
What's this?

Reply:
Copied over from the `shello.rs` example :/

Reply:
`peniko::Brush` is `PartialEq` and has been for a while so we should be able to just derive here now. I think this was a limitation in old piet-gpu code. We may need to update some revs for this to be visible.